{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2fff8c67",
   "metadata": {},
   "source": [
    "*Author: [Daniel Puente Viejo](https://www.linkedin.com/in/danielpuenteviejo/)*\n",
    "\n",
    "## **GraphRAG Tutorial — Neo4j + LLMs**\n",
    "\n",
    "<p align=\"left\">\n",
    "  <img src=\"./imgs/cover_page.gif\" alt=\"alt text\" width=\"700\"/>\n",
    "</p>\n",
    "\n",
    "Practical example of how to create a **GraphRAG** (Retrieval-Augmented Generation) system using **Neo4j** as the knowledge graph database and **OpenAI**'s language models for natural language processing tasks.   \n",
    "In this tutorial we will also make use of **LangChain** to facilitate the integration between the LLMs and the graph database.\n",
    "\n",
    "### **Index:**\n",
    "\n",
    "- <a href='#1'><ins>1. SetUp</ins></a>\n",
    "    - <a href='#1.1'><ins>1.1 Install Libraries</ins></a>\n",
    "    - <a href='#1.2'><ins>1.2 Environment Variables</ins></a>\n",
    "    - <a href='#1.3'><ins>1.3 Initialize the LLM and Embeddings</ins></a>\n",
    "    - <a href='#1.4'><ins>1.4 Initialize the graph</ins></a>\n",
    "    - <a href='#1.5'><ins>1.5 Load the chunks</ins></a>\n",
    "- <a href='#2'><ins>2. Graph construction</ins></a>\n",
    "    - <a href='#2.1'><ins>2.1 Create nodes</ins></a>\n",
    "    - <a href='#2.2'><ins>2.2 Create relationships</ins></a>\n",
    "- <a href='#3'><ins>3. Retriever</ins></a>\n",
    "    - <a href='#3.1'><ins>3.1 Define the retriever</ins></a>\n",
    "    - <a href='#3.2'><ins>3.2 Test the retriever</ins></a>\n",
    "- <a href='#4'><ins>4. Extra commands</ins></a>\n",
    "    - <a href='#4.1'><ins>4.1 Delete all nodes</ins></a>\n",
    "    - <a href='#4.2'><ins>4.2 Close the session</ins></a>\n",
    "    - <a href='#4.3'><ins>4.3 See the avaiable indexes in the Neo4j database</ins></a>\n",
    "    - <a href='#4.4'><ins>4.4 See all the filenames</ins></a>\n",
    "    - <a href='#4.5'><ins>4.5 How to filter a concrete id</ins></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b06304c",
   "metadata": {},
   "source": [
    "## <a id='1' style=\"color: skyblue;\">**1. Setup**</a>\n",
    "\n",
    "###  <a id='1.1'>**1.1 Install Libraries**</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "03644842",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard Imports\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "import pandas as pd\n",
    "import ast\n",
    "\n",
    "# LangChain Imports\n",
    "from langchain.schema import Document\n",
    "from langchain_openai import AzureChatOpenAI\n",
    "from langchain_community.graphs.neo4j_graph import Neo4jGraph\n",
    "from langchain_community.vectorstores.neo4j_vector import Neo4jVector\n",
    "from langchain_experimental.graph_transformers import LLMGraphTransformer\n",
    "from langchain_huggingface import HuggingFaceEmbeddings\n",
    "\n",
    "# Neo4j Imports\n",
    "from neo4j import GraphDatabase\n",
    "from neo4j_graphrag.retrievers import VectorRetriever\n",
    "from neo4j_graphrag.retrievers.hybrid import HybridRetriever\n",
    "from neo4j_graphrag.embeddings.sentence_transformers import SentenceTransformerEmbeddings\n",
    "from neo4j_graphrag.llm import AzureOpenAILLM\n",
    "from neo4j_graphrag.generation import GraphRAG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a2a7d0",
   "metadata": {},
   "source": [
    "### <a id='1.2'>**1.2 Environment Variables**</a>\n",
    "\n",
    "Here we load the environment variables from a .env file. Make sure to have your OpenAI API key and Neo4j credentials stored in the .env file.   \n",
    "\n",
    "*Important:* In case of any connection issue with Neo4j, try changing the `NEO4J_URI` from `neo4j+s://` to `neo4j+ssc://`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f656a6ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dotenv()\n",
    "os.environ[\"OPENAI_API_KEY\"] = os.getenv('OPENAI_KEY')\n",
    "os.environ[\"NEO4J_URI\"] = os.getenv('NEO4J_URI') # I case of any issue, change neo4j+s to neo4j+ssc\n",
    "os.environ[\"NEO4J_USERNAME\"] = os.getenv('NEO4J_USERNAME')\n",
    "os.environ[\"NEO4J_PASSWORD\"] = os.getenv('NEO4J_PASSWORD')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af02ed4",
   "metadata": {},
   "source": [
    "### <a id='1.3'>**1.3 Initialize the LLM and Embeddings**</a>\n",
    "\n",
    "For this project, we will use OpenAI's `gpt-4o-mini` for text generation and `sentence-transformers/all-MiniLM-L6-v2` for generating embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99820539",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/graphs/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "llm = AzureChatOpenAI(\n",
    "    model_name=\"gpt-4o-mini\",\n",
    "    api_key=os.getenv('OPENAI_KEY'),\n",
    "    api_version=os.getenv('OPENAI_VERSION'),\n",
    "    azure_endpoint=os.getenv('OPENAI_ENDPOINT'),\n",
    "    temperature=0\n",
    ")\n",
    "\n",
    "model_name = \"sentence-transformers/all-MiniLM-L6-v2\"\n",
    "model_kwargs = {'device': 'cpu'}\n",
    "encode_kwargs = {'normalize_embeddings': False}\n",
    "embeddings = HuggingFaceEmbeddings(\n",
    "    model_name=model_name,\n",
    "    model_kwargs=model_kwargs,\n",
    "    encode_kwargs=encode_kwargs\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17b24219",
   "metadata": {},
   "source": [
    "### <a id='1.4'>**1.4 Initialize the graph**</a>\n",
    "\n",
    "Here we will get the Neo4jGraph instance, the driver and the session to interact with the Neo4j database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c2a51350",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0l/4r77qdrs2yxfrd5z2d_2n89w0000gp/T/ipykernel_22064/300369370.py:1: LangChainDeprecationWarning: The class `Neo4jGraph` was deprecated in LangChain 0.3.8 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-neo4j package and should be used instead. To use it run `pip install -U :class:`~langchain-neo4j` and import as `from :class:`~langchain_neo4j import Neo4jGraph``.\n",
      "  graph = Neo4jGraph()\n"
     ]
    }
   ],
   "source": [
    "graph = Neo4jGraph()\n",
    "driver = GraphDatabase.driver(uri=os.getenv('NEO4J_URI'), auth=(os.getenv('NEO4J_USERNAME'), os.getenv('NEO4J_PASSWORD')))\n",
    "session = driver.session()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e84214a",
   "metadata": {},
   "source": [
    "### <a id='1.5'>**1.5 Load the chunks**</a>\n",
    "\n",
    "In this section, we will load the document chunks into the Neo4j graph database.\n",
    "\n",
    "In this case we already have the chunks stores in a CSV file. It is a very simple CSV file that contains all the information inside the text files on `/data/original_data/`. You can take a look to them if you want to see the original documents.   \n",
    "It is synthetic data that has been generated just for this tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8083fd75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chunk_id</th>\n",
       "      <th>text</th>\n",
       "      <th>filename</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7f8d58f9be5543c0952d60084f7eb160</td>\n",
       "      <td>During the 2024 European Society for Medical O...</td>\n",
       "      <td>oncology_update_spain.txt</td>\n",
       "      <td>{\"filename\": \"oncology_update_spain.txt\", \"cou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c006dc492ec541b8bc9ee9aa9e634d30</td>\n",
       "      <td>Another highlight came from the collaborative ...</td>\n",
       "      <td>oncology_update_spain.txt</td>\n",
       "      <td>{\"filename\": \"oncology_update_spain.txt\", \"cou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bbc171c93c5a47c1805d888489d75ea9</td>\n",
       "      <td>The Spanish Oncology Network (RedOnco) announc...</td>\n",
       "      <td>oncology_update_spain.txt</td>\n",
       "      <td>{\"filename\": \"oncology_update_spain.txt\", \"cou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>69c14a8b63da4cf3b18fabbd00cefb4f</td>\n",
       "      <td>The CARDI-RED study, conducted by the Mayo Cli...</td>\n",
       "      <td>cardio_study_usa.txt</td>\n",
       "      <td>{\"filename\": \"cardio_study_usa.txt\", \"country\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>a6a430bdc7764d3185888aa4beb3d026</td>\n",
       "      <td>A genomic substudy within CARDI-RED identified...</td>\n",
       "      <td>cardio_study_usa.txt</td>\n",
       "      <td>{\"filename\": \"cardio_study_usa.txt\", \"country\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2b501bb8caf74fb9ab814f8af9530576</td>\n",
       "      <td>A genomic substudy within CARDI-RED identified...</td>\n",
       "      <td>cardio_study_usa.txt</td>\n",
       "      <td>{\"filename\": \"cardio_study_usa.txt\", \"country\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>c80f7c811e8d485e832a445f6199f967</td>\n",
       "      <td>The IL-6 Inhibition Trial (UK-IL6-2024) was a ...</td>\n",
       "      <td>trial_inflammation_uk.txt</td>\n",
       "      <td>{\"filename\": \"trial_inflammation_uk.txt\", \"cou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cda3cfe5ae894b11bcbad195cee4a221</td>\n",
       "      <td>A post-hoc biomarker analysis performed by the...</td>\n",
       "      <td>trial_inflammation_uk.txt</td>\n",
       "      <td>{\"filename\": \"trial_inflammation_uk.txt\", \"cou...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           chunk_id  \\\n",
       "0  7f8d58f9be5543c0952d60084f7eb160   \n",
       "1  c006dc492ec541b8bc9ee9aa9e634d30   \n",
       "2  bbc171c93c5a47c1805d888489d75ea9   \n",
       "3  69c14a8b63da4cf3b18fabbd00cefb4f   \n",
       "4  a6a430bdc7764d3185888aa4beb3d026   \n",
       "5  2b501bb8caf74fb9ab814f8af9530576   \n",
       "6  c80f7c811e8d485e832a445f6199f967   \n",
       "7  cda3cfe5ae894b11bcbad195cee4a221   \n",
       "\n",
       "                                                text  \\\n",
       "0  During the 2024 European Society for Medical O...   \n",
       "1  Another highlight came from the collaborative ...   \n",
       "2  The Spanish Oncology Network (RedOnco) announc...   \n",
       "3  The CARDI-RED study, conducted by the Mayo Cli...   \n",
       "4  A genomic substudy within CARDI-RED identified...   \n",
       "5  A genomic substudy within CARDI-RED identified...   \n",
       "6  The IL-6 Inhibition Trial (UK-IL6-2024) was a ...   \n",
       "7  A post-hoc biomarker analysis performed by the...   \n",
       "\n",
       "                    filename  \\\n",
       "0  oncology_update_spain.txt   \n",
       "1  oncology_update_spain.txt   \n",
       "2  oncology_update_spain.txt   \n",
       "3       cardio_study_usa.txt   \n",
       "4       cardio_study_usa.txt   \n",
       "5       cardio_study_usa.txt   \n",
       "6  trial_inflammation_uk.txt   \n",
       "7  trial_inflammation_uk.txt   \n",
       "\n",
       "                                            metadata  \n",
       "0  {\"filename\": \"oncology_update_spain.txt\", \"cou...  \n",
       "1  {\"filename\": \"oncology_update_spain.txt\", \"cou...  \n",
       "2  {\"filename\": \"oncology_update_spain.txt\", \"cou...  \n",
       "3  {\"filename\": \"cardio_study_usa.txt\", \"country\"...  \n",
       "4  {\"filename\": \"cardio_study_usa.txt\", \"country\"...  \n",
       "5  {\"filename\": \"cardio_study_usa.txt\", \"country\"...  \n",
       "6  {\"filename\": \"trial_inflammation_uk.txt\", \"cou...  \n",
       "7  {\"filename\": \"trial_inflammation_uk.txt\", \"cou...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PATH = \"data/documents.csv\"\n",
    "\n",
    "df = pd.read_csv(PATH)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e08bfc23",
   "metadata": {},
   "source": [
    "## <a id='2' style=\"color: skyblue;\">**2. Graph construction**</a>\n",
    "\n",
    "### <a id='2.1'>**2.1 Adapt the format**</a>\n",
    "\n",
    "Neo4jGraph requires a specific format to ingest data. We will adapt the format of our CSV file to match the required structure.   \n",
    "We need to use `Document` from `langchain.schema` to create the documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8bf001a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'filename': 'oncology_update_spain.txt', 'country': 'Spain', 'topic': 'Oncology', 'source_type': 'Congress Summary', 'page_number': 1, 'title': 'ESMO 2024: Advances in EGFR Inhibition', 'keywords': ['AZD5478', 'EGFR', 'NSCLC', 'AstraZeneca', 'ESMO'], 'main_investigator': 'Dr. Isabel Romero', 'institution': 'Hospital Vall d’Hebron', 'year': 2024}, page_content='During the 2024 European Society for Medical Oncology (ESMO) Congress in Madrid, Spanish oncology researchers presented a series of groundbreaking updates on targeted and immunotherapy-based treatments. The centerpiece of the conference was AZD5478, a next-generation EGFR inhibitor developed by AstraZeneca, which demonstrated significant activity in non-small-cell lung cancer (NSCLC) patients harboring EGFR exon 20 insertions. The Phase II data showed an objective response rate of 56% and median progression-free survival (PFS) of 10.4 months, surpassing current standards of care.\\n\\nDr. Isabel Romero from Hospital Vall d’Hebron in Barcelona highlighted the importance of molecular profiling, noting that “precision oncology is no longer an aspiration—it is the standard.” The presentation emphasized the need for routine next-generation sequencing (NGS) to detect resistance mutations and guide therapy selection. Moreover, safety outcomes were favorable, with diarrhea and rash being the most frequent adverse events, consistent with other EGFR inhibitors.\\n\\nThis presentation marked a shift toward comprehensive mutation testing across all lung cancer patients in Spain, supported by new reimbursement policies from the Ministry of Health that incentivize genomic diagnostics in tertiary hospitals.'),\n",
       " Document(metadata={'filename': 'oncology_update_spain.txt', 'country': 'Spain', 'topic': 'Oncology', 'source_type': 'Congress Summary', 'page_number': 2, 'title': 'Combination Therapies in Triple-Negative Breast Cancer', 'keywords': ['Pembrolizumab', 'Sacituzumab Govitecan', 'TNBC', 'GEICAM', 'Immunotherapy'], 'main_investigator': 'Dr. Laura Sánchez', 'institution': 'Hospital 12 de Octubre', 'year': 2024}, page_content='Another highlight came from the collaborative trial between Hospital 12 de Octubre (Madrid) and the Spanish Breast Cancer Group (GEICAM), focusing on combination therapies for triple-negative breast cancer (TNBC). The study evaluated pembrolizumab in combination with the antibody-drug conjugate sacituzumab govitecan in patients with metastatic disease who had received at least two prior lines of therapy. The combination achieved an overall response rate of 28% and disease control rate of 64%, with median overall survival extending to 13.2 months.\\n\\nResearchers also discussed biomarker-driven insights, revealing that patients with PD-L1 expression ≥10% and BRCA1/2 mutations had significantly higher response rates. “These findings underscore the evolving role of immunotherapy even in traditionally hard-to-treat tumor types,” stated Dr. Laura Sánchez, co-investigator and immuno-oncologist at Hospital 12 de Octubre. Safety data indicated manageable toxicity, with neutropenia and diarrhea as the leading adverse events.\\n\\nThe symposium closed with a call for greater collaboration between molecular pathology laboratories and oncology units to accelerate patient stratification and trial enrollment.'),\n",
       " Document(metadata={'filename': 'oncology_update_spain.txt', 'country': 'Spain', 'topic': 'Oncology', 'source_type': 'Congress Summary', 'page_number': 3, 'title': 'The OncoDataLink Initiative', 'keywords': ['Data Integration', 'OncoDataLink', 'Genomics', 'AI', 'Real-World Evidence'], 'main_investigator': 'Dr. Isabel Romero', 'institution': 'RedOnco Network', 'year': 2024}, page_content='The Spanish Oncology Network (RedOnco) announced the launch of \"OncoDataLink,\" a national digital infrastructure designed to integrate clinical, genomic, and radiomic data from over 30 cancer research centers across Spain. The initiative, supported by European funding, aims to harmonize datasets, enable AI-driven pattern recognition, and optimize clinical trial recruitment through federated learning models.\\n\\nDr. Romero emphasized that such integration will “transform how oncologists interpret data, allowing faster translation of molecular insights into bedside practice.” The network will also collaborate with AstraZeneca’s Real-World Evidence division to develop dashboards for continuous post-market surveillance of novel cancer therapeutics.\\n\\nThe Congress concluded that the next frontier in oncology will hinge not only on discovering new molecules but also on data harmonization and computational intelligence to uncover unseen relationships between drugs, genes, and patient outcomes.'),\n",
       " Document(metadata={'filename': 'cardio_study_usa.txt', 'country': 'USA', 'topic': 'Cardiology', 'source_type': 'Publication Abstracts', 'page_number': 1, 'title': 'CARDI-RED: Non-Statin LDL Reduction Strategies', 'keywords': ['Bempedoic Acid', 'Ezetimibe', 'LDL', 'Mayo Clinic', 'Cleveland Heart Center'], 'main_investigator': 'Dr. Michael Nguyen', 'institution': 'Mayo Clinic', 'year': 2024}, page_content='The CARDI-RED study, conducted by the Mayo Clinic in collaboration with the Cleveland Heart Center, explored lipid-lowering strategies for patients resistant or intolerant to statins. This multicenter, double-blind trial evaluated the efficacy of a fixed-dose combination of bempedoic acid and ezetimibe compared with placebo in 310 patients with LDL cholesterol (LDL-C) levels above 160 mg/dL despite lifestyle modification and previous statin therapy.\\n\\nAfter 24 weeks, patients receiving the combination therapy demonstrated a mean LDL-C reduction of 34% relative to baseline, compared with a 5% reduction in the placebo arm (p<0.001). Importantly, no significant increase in muscle-related adverse events, commonly associated with statins, was observed. High-sensitivity C-reactive protein (hs-CRP) levels also decreased by 23%, indicating an anti-inflammatory benefit beyond lipid lowering.\\n\\n“These findings could redefine the standard of care for patients who cannot tolerate statins,” commented Dr. Michael Nguyen, principal investigator of the trial. “Our data show that bempedoic acid provides a safe, oral alternative for long-term cardiovascular risk management.”'),\n",
       " Document(metadata={'filename': 'cardio_study_usa.txt', 'country': 'USA', 'topic': 'Cardiology', 'source_type': 'Publication Abstracts', 'page_number': 2, 'title': 'Pharmacogenomic Insights and Future Directions', 'keywords': ['PCSK9', 'Genotyping', 'Pharmacogenomics', 'FDA', 'Cardiovascular Risk'], 'main_investigator': 'Dr. Michael Nguyen', 'institution': 'Mayo Clinic', 'year': 2024}, page_content='A genomic substudy within CARDI-RED identified a subgroup of 62 patients carrying PCSK9 gene variants who exhibited even greater LDL-C reduction, averaging 42%. This observation suggests a potential gene-drug interaction, reinforcing the growing relevance of pharmacogenomics in cardiovascular prevention.\\n\\nThe study team recommended that future guidelines include genotyping for PCSK9 and SLCO1B1 polymorphisms to personalize therapy. The authors also noted that combining bempedoic acid with PCSK9 monoclonal antibodies could yield additive effects, paving the way for combination strategies in refractory hypercholesterolemia.\\n\\nThe research consortium plans to extend the study to evaluate long-term outcomes such as myocardial infarction rates and cardiovascular mortality. The U.S. Food and Drug Administration (FDA) has expressed interest in reviewing these findings as part of its ongoing assessment of non-statin lipid-lowering therapies.'),\n",
       " Document(metadata={'filename': 'cardio_study_usa.txt', 'country': 'USA', 'topic': 'Cardiology', 'source_type': 'Publication Abstracts'}, page_content='A genomic substudy within CARDI-RED identified a subgroup of 62 patients carrying PCSK9 gene variants who exhibited even greater LDL-C reduction, averaging 42%. This observation suggests a potential gene-drug interaction, reinforcing the growing relevance of pharmacogenomics in cardiovascular prevention.\\n\\nThe study team recommended that future guidelines include genotyping for PCSK9 and SLCO1B1 polymorphisms to personalize therapy. The authors also noted that combining bempedoic acid with PCSK9 monoclonal antibodies could yield additive effects, paving the way for combination strategies in refractory hypercholesterolemia.\\n\\nThe research consortium plans to extend the study to evaluate long-term outcomes such as myocardial infarction rates and cardiovascular mortality. The U.S. Food and Drug Administration (FDA) has expressed interest in reviewing these findings as part of its ongoing assessment of non-statin lipid-lowering therapies.'),\n",
       " Document(metadata={'filename': 'trial_inflammation_uk.txt', 'country': 'United Kingdom', 'topic': 'Inflammatory Diseases', 'source_type': 'Clinical Trial Report', 'page_number': 1, 'title': 'IL-6 Inhibition Trial – Phase II Study', 'keywords': ['Liximab', 'IL-6', 'Rheumatoid Arthritis', 'Inflammation', 'Oxford'], 'main_investigator': 'Dr. Amelia Wright', 'institution': 'University of Oxford', 'year': 2024}, page_content='The IL-6 Inhibition Trial (UK-IL6-2024) was a randomized, double-blind, placebo-controlled Phase II study evaluating the efficacy and safety of the monoclonal antibody Liximab in patients with moderate to severe rheumatoid arthritis. The trial, coordinated by the University of Oxford and conducted across five UK centers (London, Manchester, Edinburgh, Bristol, and Leeds), enrolled 240 adults aged between 35 and 70 years. Participants had an inadequate response to methotrexate and were randomized 2:1 to receive either Liximab or placebo, both administered via subcutaneous injection every two weeks.\\n\\nAfter 12 weeks of treatment, patients in the Liximab group demonstrated a statistically significant improvement in the American College of Rheumatology 50 (ACR50) response rate, reaching 62% compared with 28% in the placebo group. Reductions in C-reactive protein (CRP) and erythrocyte sedimentation rate (ESR) levels were observed as early as week 4. Adverse events were generally mild to moderate in intensity, with fatigue, nausea, and injection-site reactions being the most common. Two cases of transient liver enzyme elevation were reported, both resolving spontaneously without medical intervention.\\n\\nThe investigators emphasized that Liximab’s mechanism of action—targeting the IL-6 receptor—aligns with established pathways involved in chronic inflammation and joint destruction. No cases of serious infection were detected, and overall tolerability was consistent with the safety profile of existing IL-6 inhibitors.'),\n",
       " Document(metadata={'filename': 'trial_inflammation_uk.txt', 'country': 'United Kingdom', 'topic': 'Inflammatory Diseases', 'source_type': 'Clinical Trial Report', 'page_number': 2, 'title': 'Biomarker Analysis and Future Directions', 'keywords': ['Biomarkers', 'IL-6', 'TNF-alpha', 'AstraZeneca', 'UK-MHRA'], 'main_investigator': 'Dr. Amelia Wright', 'institution': 'University of Oxford', 'year': 2024}, page_content='A post-hoc biomarker analysis performed by the Oxford Rheumatology Research Group revealed that treatment with Liximab significantly reduced circulating levels of interleukin-6 (IL-6) and tumor necrosis factor-alpha (TNF-α), confirming effective pathway inhibition. The reduction in serum amyloid A and fibrinogen concentrations further supported a systemic anti-inflammatory effect. Imaging data from ultrasound scans showed a marked reduction in synovial vascularity after 8 weeks of therapy.\\n\\nThe study was led by Dr. Amelia Wright, principal investigator and senior lecturer at the University of Oxford. Collaborators included the National Rheumatoid Arthritis Society and funding support from AstraZeneca’s Early Immunology program. Dr. Wright stated that “Liximab demonstrates a compelling efficacy signal, suggesting potential disease modification when used early in the course of rheumatoid arthritis.”\\n\\nA follow-up Phase III trial is being designed to assess long-term safety, radiographic progression, and comparative efficacy against Tocilizumab and Adalimumab. Discussions with the UK Medicines and Healthcare Regulatory Agency (MHRA) are ongoing to define optimal endpoints and patient stratification criteria.')]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "documents = [\n",
    "    Document(\n",
    "        page_content=str(element['text']),\n",
    "        metadata=ast.literal_eval(element['metadata'])\n",
    "    )\n",
    "    for _, element in df.iterrows()\n",
    "]\n",
    "\n",
    "documents"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07d4bafc",
   "metadata": {},
   "source": [
    "### <a id='2.2'>**2.2 Identify the nodes and relationships**</a>\n",
    "\n",
    "Now, we will identify the nodes and relationships in our documents using LangChain's `Neo4jGraph` capabilities.   \n",
    "The nodes and relationships used in this example are dummy and for demonstration purposes only. You can modify them according to your needs.\n",
    "\n",
    "* Nodes: Nodes represent entities or concepts in the graph.\n",
    "* Relationships: Relationships define how nodes are connected. We will create relationships such as `WORKS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5d4a2fcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "allowed_nodes = [\n",
    "    \"Drug\", \"Disease\", \"Biomarker\", \"Trial\",\n",
    "    \"Researcher\", \"Institution\", \"Company\", \"Country\"\n",
    "]\n",
    "\n",
    "allowed_relationships = [\n",
    "    \"TREATS\", \"TARGETS\", \"EVALUATED_IN\", \"STUDIES\",\n",
    "    \"CONDUCTED_BY\", \"LEAD_BY\", \"SPONSORED_BY\", \"LOCATED_IN\"\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "376ede84",
   "metadata": {},
   "source": [
    "### <a id='2.3'>**2.3 LLM to identify the nodes and relationships**</a>\n",
    "\n",
    "To accurately identify nodes and relationships from unstructured text, we will leverage the capabilities of our LLM. The model will analyze the content of each document chunk and extract relevant entities and their connections based on predefined criteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6f542365",
   "metadata": {},
   "outputs": [],
   "source": [
    "transformer = LLMGraphTransformer(\n",
    "    llm=llm,\n",
    "    allowed_nodes=allowed_nodes,\n",
    "    allowed_relationships=allowed_relationships,\n",
    "    node_properties=False, \n",
    "    relationship_properties=False\n",
    ") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4e21b8c",
   "metadata": {},
   "source": [
    "### <a id='2.4'>**2.4 Add the nodes**</a>\n",
    "\n",
    "Now that we have identified the nodes and relationships, we will proceed to add them to the Neo4j graph database using LangChain's `Neo4jGraph` functionalities.\n",
    "Thanks to the previous `LLMGraphTransformer`, the nodes and relationships will be automatically created in the graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6602ac3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_documents = transformer.convert_to_graph_documents(documents)\n",
    "graph.add_graph_documents(graph_documents, include_source=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "463f01c0",
   "metadata": {},
   "source": [
    "### <a id='2.5'>**2.5 Create the keyword index**</a>\n",
    "\n",
    "The keyword index is very important as will help to create hybrid searches combining both vector and keyword-based searches, alongside with the graph structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9d69837c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<neo4j._sync.work.result.Result at 0x3088d49e0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session.run(\"DROP INDEX keyword IF EXISTS\")\n",
    "session.run(\"\"\"\n",
    "CREATE FULLTEXT INDEX keyword FOR (n:Document) ON EACH [n.text]\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016b98b6",
   "metadata": {},
   "source": [
    "### <a id='2.6'>**2.6 Add the embeddings**</a>\n",
    "\n",
    "Now we will add the embeddings to the nodes in the graph. This will allow us to perform vector-based similarity searches later on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f15193e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = Neo4jVector.from_documents(\n",
    "    documents,\n",
    "    embeddings,\n",
    "    pre_delete_collection=True, \n",
    "    index_name=\"vector\",\n",
    "    keyword_index_name=\"keyword\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8617d101",
   "metadata": {},
   "source": [
    "## <a id='3' style=\"color: skyblue;\">**3. Retriever**</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f83f1a06",
   "metadata": {},
   "source": [
    "### <a id='3.1'>**3.1. Vector retriever (Clasical RAG)**</a>\n",
    "\n",
    "This is an example of how to generate a vector retriever using the Neo4jGrag library.  \n",
    "The behavior is like a RAG system, where the retriever fetches relevant nodes based on vector similarity, it does not consider relationships between nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f9bbf5c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "embedder = SentenceTransformerEmbeddings()\n",
    "\n",
    "vector_retriever = VectorRetriever(\n",
    "    driver,\n",
    "    embedder=embedder,\n",
    "    index_name=\"vector\",\n",
    "    return_properties=[\"text\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6cf8067",
   "metadata": {},
   "source": [
    "#### **3.1.1 Testing the Vector Retriever**\n",
    "\n",
    "In this example, we will test the vector retriever by providing a query and retrieving the top relevant nodes based on vector similarity.    \n",
    "We will display the first result to see the details of the most relevant chunk."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cc47c405",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'content': \"{'text': 'Another highlight came from the collaborative trial between Hospital 12 de Octubre (Madrid) and the Spanish Breast Cancer Group (GEICAM), focusing on combination therapies for triple-negative breast cancer (TNBC). The study evaluated pembrolizumab in combination with the antibody-drug conjugate sacituzumab govitecan in patients with metastatic disease who had received at least two prior lines of therapy. The combination achieved an overall response rate of 28% and disease control rate of 64%, with median overall survival extending to 13.2 months.\\\\n\\\\nResearchers also discussed biomarker-driven insights, revealing that patients with PD-L1 expression ≥10% and BRCA1/2 mutations had significantly higher response rates. “These findings underscore the evolving role of immunotherapy even in traditionally hard-to-treat tumor types,” stated Dr. Laura Sánchez, co-investigator and immuno-oncologist at Hospital 12 de Octubre. Safety data indicated manageable toxicity, with neutropenia and diarrhea as the leading adverse events.\\\\n\\\\nThe symposium closed with a call for greater collaboration between molecular pathology laboratories and oncology units to accelerate patient stratification and trial enrollment.'}\",\n",
       "  'metadata': {'score': 0.7780308723449707,\n",
       "   'nodeLabels': ['Chunk'],\n",
       "   'id': '4:50a55e0e-0d46-4a3c-85dd-d974565f398f:41'}},\n",
       " {'content': \"{'text': 'During the 2024 European Society for Medical Oncology (ESMO) Congress in Madrid, Spanish oncology researchers presented a series of groundbreaking updates on targeted and immunotherapy-based treatments. The centerpiece of the conference was AZD5478, a next-generation EGFR inhibitor developed by AstraZeneca, which demonstrated significant activity in non-small-cell lung cancer (NSCLC) patients harboring EGFR exon 20 insertions. The Phase II data showed an objective response rate of 56% and median progression-free survival (PFS) of 10.4 months, surpassing current standards of care.\\\\n\\\\nDr. Isabel Romero from Hospital Vall d’Hebron in Barcelona highlighted the importance of molecular profiling, noting that “precision oncology is no longer an aspiration—it is the standard.” The presentation emphasized the need for routine next-generation sequencing (NGS) to detect resistance mutations and guide therapy selection. Moreover, safety outcomes were favorable, with diarrhea and rash being the most frequent adverse events, consistent with other EGFR inhibitors.\\\\n\\\\nThis presentation marked a shift toward comprehensive mutation testing across all lung cancer patients in Spain, supported by new reimbursement policies from the Ministry of Health that incentivize genomic diagnostics in tertiary hospitals.'}\",\n",
       "  'metadata': {'score': 0.7266240119934082,\n",
       "   'nodeLabels': ['Chunk'],\n",
       "   'id': '4:50a55e0e-0d46-4a3c-85dd-d974565f398f:40'}}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"Which researchers in Spain are involved in studies evaluating immunotherapy drugs for breast cancer?\"\n",
    "\n",
    "retriever_result = vector_retriever.search(query_text=query, top_k=2)\n",
    "\n",
    "# Example of the first result\n",
    "ast.literal_eval(retriever_result.model_dump_json())['items']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bdca83c",
   "metadata": {},
   "source": [
    "### <a id='3.2'>**3.2 Hybrid retriever (Graph RAG)**</a>\n",
    "\n",
    "This is an example of how to generate a hybrid retriever using the Neo4jGrag library.  \n",
    "The behavior is like a Graph RAG system, where the retriever fetches relevant nodes based on both vector similarity and graph relationships."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6137a7f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "embedder = SentenceTransformerEmbeddings()\n",
    "\n",
    "graphrag_retriever = HybridRetriever(\n",
    "    driver,\n",
    "    embedder=embedder,\n",
    "    vector_index_name=\"vector\",\n",
    "    fulltext_index_name=\"keyword\",\n",
    "    return_properties=[\"text\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b90ad76",
   "metadata": {},
   "source": [
    "#### **3.2.1 Testing the Graph Retriever**\n",
    "\n",
    "In this example, we will test the graph retriever by providing a query and retrieving the top relevant nodes based on vector similarity and graph relationships."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2ec3bd69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'content': \"{'text': 'Another highlight came from the collaborative trial between Hospital 12 de Octubre (Madrid) and the Spanish Breast Cancer Group (GEICAM), focusing on combination therapies for triple-negative breast cancer (TNBC). The study evaluated pembrolizumab in combination with the antibody-drug conjugate sacituzumab govitecan in patients with metastatic disease who had received at least two prior lines of therapy. The combination achieved an overall response rate of 28% and disease control rate of 64%, with median overall survival extending to 13.2 months.\\\\n\\\\nResearchers also discussed biomarker-driven insights, revealing that patients with PD-L1 expression ≥10% and BRCA1/2 mutations had significantly higher response rates. “These findings underscore the evolving role of immunotherapy even in traditionally hard-to-treat tumor types,” stated Dr. Laura Sánchez, co-investigator and immuno-oncologist at Hospital 12 de Octubre. Safety data indicated manageable toxicity, with neutropenia and diarrhea as the leading adverse events.\\\\n\\\\nThe symposium closed with a call for greater collaboration between molecular pathology laboratories and oncology units to accelerate patient stratification and trial enrollment.'}\",\n",
       "  'metadata': {'score': 1.0}},\n",
       " {'content': \"{'text': 'During the 2024 European Society for Medical Oncology (ESMO) Congress in Madrid, Spanish oncology researchers presented a series of groundbreaking updates on targeted and immunotherapy-based treatments. The centerpiece of the conference was AZD5478, a next-generation EGFR inhibitor developed by AstraZeneca, which demonstrated significant activity in non-small-cell lung cancer (NSCLC) patients harboring EGFR exon 20 insertions. The Phase II data showed an objective response rate of 56% and median progression-free survival (PFS) of 10.4 months, surpassing current standards of care.\\\\n\\\\nDr. Isabel Romero from Hospital Vall d’Hebron in Barcelona highlighted the importance of molecular profiling, noting that “precision oncology is no longer an aspiration—it is the standard.” The presentation emphasized the need for routine next-generation sequencing (NGS) to detect resistance mutations and guide therapy selection. Moreover, safety outcomes were favorable, with diarrhea and rash being the most frequent adverse events, consistent with other EGFR inhibitors.\\\\n\\\\nThis presentation marked a shift toward comprehensive mutation testing across all lung cancer patients in Spain, supported by new reimbursement policies from the Ministry of Health that incentivize genomic diagnostics in tertiary hospitals.'}\",\n",
       "  'metadata': {'score': 1.0}}]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"Which researchers in Spain are involved in studies evaluating immunotherapy drugs for breast cancer?\"\n",
    "\n",
    "retriever_result = graphrag_retriever.search(query_text=query, top_k=2)\n",
    "\n",
    "# Example of the first result\n",
    "ast.literal_eval(retriever_result.model_dump_json())['items']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf13d6b8",
   "metadata": {},
   "source": [
    "#### **3.2.2 Generate Graph RAG retriever-and-answer**\n",
    "\n",
    "We will create a hybrid retriever that combines vector similarity and graph relationships to fetch relevant nodes from the Neo4j graph database, but incorporating the LLM inside so we automatically return the answer apart from the relevant chunks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b4501f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm_graphrag = AzureOpenAILLM(\n",
    "    model_name=\"gpt-4o-mini\",\n",
    "    api_key=os.getenv('OPENAI_KEY'),\n",
    "    api_version=os.getenv('OPENAI_VERSION'),\n",
    "    azure_endpoint=os.getenv('OPENAI_ENDPOINT'),\n",
    ")\n",
    "\n",
    "rag = GraphRAG(\n",
    "    retriever=graphrag_retriever, \n",
    "    llm=llm_graphrag,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e813d300",
   "metadata": {},
   "source": [
    "#### **3.2.3 Testing the Graph RAG Retriever + Generation**\n",
    "\n",
    "Now we can test the Graph RAG retriever by providing a query.   \n",
    "As you can see the retriever configuration has not the exact same format as the vector retriever. However, they mantain similar parameters such as `top_k`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1d4b7b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"Which researchers in Spain are involved in studies evaluating immunotherapy drugs for breast cancer?\"\n",
    "\n",
    "response = rag.search(\n",
    "    query_text=query,\n",
    "    retriever_config={\n",
    "        \"top_k\": 2,\n",
    "        # \"filters\": {}\n",
    "    },\n",
    "    return_context=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f609a8fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'content': \"{'text': 'Another highlight came from the collaborative trial between Hospital 12 de Octubre (Madrid) and the Spanish Breast Cancer Group (GEICAM), focusing on combination therapies for triple-negative breast cancer (TNBC). The study evaluated pembrolizumab in combination with the antibody-drug conjugate sacituzumab govitecan in patients with metastatic disease who had received at least two prior lines of therapy. The combination achieved an overall response rate of 28% and disease control rate of 64%, with median overall survival extending to 13.2 months.\\\\n\\\\nResearchers also discussed biomarker-driven insights, revealing that patients with PD-L1 expression ≥10% and BRCA1/2 mutations had significantly higher response rates. “These findings underscore the evolving role of immunotherapy even in traditionally hard-to-treat tumor types,” stated Dr. Laura Sánchez, co-investigator and immuno-oncologist at Hospital 12 de Octubre. Safety data indicated manageable toxicity, with neutropenia and diarrhea as the leading adverse events.\\\\n\\\\nThe symposium closed with a call for greater collaboration between molecular pathology laboratories and oncology units to accelerate patient stratification and trial enrollment.'}\",\n",
       "  'metadata': {'score': 1.0}},\n",
       " {'content': \"{'text': 'During the 2024 European Society for Medical Oncology (ESMO) Congress in Madrid, Spanish oncology researchers presented a series of groundbreaking updates on targeted and immunotherapy-based treatments. The centerpiece of the conference was AZD5478, a next-generation EGFR inhibitor developed by AstraZeneca, which demonstrated significant activity in non-small-cell lung cancer (NSCLC) patients harboring EGFR exon 20 insertions. The Phase II data showed an objective response rate of 56% and median progression-free survival (PFS) of 10.4 months, surpassing current standards of care.\\\\n\\\\nDr. Isabel Romero from Hospital Vall d’Hebron in Barcelona highlighted the importance of molecular profiling, noting that “precision oncology is no longer an aspiration—it is the standard.” The presentation emphasized the need for routine next-generation sequencing (NGS) to detect resistance mutations and guide therapy selection. Moreover, safety outcomes were favorable, with diarrhea and rash being the most frequent adverse events, consistent with other EGFR inhibitors.\\\\n\\\\nThis presentation marked a shift toward comprehensive mutation testing across all lung cancer patients in Spain, supported by new reimbursement policies from the Ministry of Health that incentivize genomic diagnostics in tertiary hospitals.'}\",\n",
       "  'metadata': {'score': 1.0}}]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.retriever_result.model_dump()['items']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "da1ac7e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Researchers from Hospital 12 de Octubre in Madrid, specifically Dr. Laura Sánchez, are involved in studies evaluating immunotherapy drugs for breast cancer, particularly in a trial focusing on combination therapies for triple-negative breast cancer (TNBC).\n"
     ]
    }
   ],
   "source": [
    "print(response.answer)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d265a23",
   "metadata": {},
   "source": [
    "## <a id='4' style=\"color: skyblue;\">**4. Extra commands**</a>\n",
    "\n",
    "In this section, we provide some extra commands that can be useful for managing and interacting with the Neo4j graph database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27a3016d",
   "metadata": {},
   "source": [
    "### <a id='4.1'>**4.1 Delete all nodes**</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d76608f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with driver.session() as session:\n",
    "#     session.run(\"MATCH (n) DETACH DELETE n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f89243c1",
   "metadata": {},
   "source": [
    "### <a id='4.2'>**4.2 Close the session**</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6989fd3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# session.close()\n",
    "# driver.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b8bd4ad",
   "metadata": {},
   "source": [
    "### <a id='4.3'>**4.3 See the avaiable indexes in the Neo4j database**</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5affc37d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Record id=3 name='constraint_f8c8c4e0' state='ONLINE' populationPercent=100.0 type='RANGE' entityType='NODE' labelsOrTypes=['Chunk'] properties=['id'] indexProvider='range-1.0' owningConstraint='constraint_f8c8c4e0' lastRead=neo4j.time.DateTime(2025, 11, 10, 11, 50, 28, 698000000, tzinfo=<UTC>) readCount=88>\n",
      "<Record id=1 name='index_1b9dcc97' state='ONLINE' populationPercent=100.0 type='LOOKUP' entityType='RELATIONSHIP' labelsOrTypes=None properties=None indexProvider='token-lookup-1.0' owningConstraint=None lastRead=neo4j.time.DateTime(2025, 11, 10, 12, 8, 13, 147000000, tzinfo=<UTC>) readCount=3>\n",
      "<Record id=0 name='index_460996c0' state='ONLINE' populationPercent=100.0 type='LOOKUP' entityType='NODE' labelsOrTypes=None properties=None indexProvider='token-lookup-1.0' owningConstraint=None lastRead=neo4j.time.DateTime(2025, 11, 10, 12, 7, 50, 896000000, tzinfo=<UTC>) readCount=646>\n",
      "<Record id=5 name='keyword' state='ONLINE' populationPercent=100.0 type='FULLTEXT' entityType='NODE' labelsOrTypes=['Document'] properties=['text'] indexProvider='fulltext-2.0' owningConstraint=None lastRead=neo4j.time.DateTime(2025, 11, 10, 11, 50, 35, 816000000, tzinfo=<UTC>) readCount=1>\n",
      "<Record id=2 name='vector' state='ONLINE' populationPercent=100.0 type='VECTOR' entityType='NODE' labelsOrTypes=['Chunk'] properties=['embedding'] indexProvider='vector-3.0' owningConstraint=None lastRead=neo4j.time.DateTime(2025, 11, 10, 11, 50, 35, 800000000, tzinfo=<UTC>) readCount=2>\n"
     ]
    }
   ],
   "source": [
    "query = \"SHOW INDEXES;\"\n",
    "with driver.session() as session:\n",
    "    result = session.run(query)\n",
    "    for record in result:\n",
    "        print(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26e0f204",
   "metadata": {},
   "source": [
    "### <a id='4.4'>**4.4 See all the filenames**</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c9f3fc94",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cardio_study_usa.txt\n",
      "oncology_update_spain.txt\n",
      "trial_inflammation_uk.txt\n"
     ]
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (n)\n",
    "WHERE n.filename IS NOT NULL\n",
    "RETURN DISTINCT n.filename AS filename\n",
    "ORDER BY filename\n",
    "\"\"\"\n",
    "\n",
    "with driver.session() as session:\n",
    "    for record in session.run(query):\n",
    "        print(record[\"filename\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f517071",
   "metadata": {},
   "source": [
    "### <a id='4.5'>**4.5 How to filter a concrete id**</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6c2bb697",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "During the 2024 European Society for Medical Oncology (ESMO) Congress in Madrid, Spanish oncology researchers presented a series of groundbreaking updates on targeted and immunotherapy-based treatments. The centerpiece of the conference was AZD5478, a next-generation EGFR inhibitor developed by AstraZeneca, which demonstrated significant activity in non-small-cell lung cancer (NSCLC) patients harboring EGFR exon 20 insertions. The Phase II data showed an objective response rate of 56% and median progression-free survival (PFS) of 10.4 months, surpassing current standards of care.\n",
      "\n",
      "Dr. Isabel Romero from Hospital Vall d’Hebron in Barcelona highlighted the importance of molecular profiling, noting that “precision oncology is no longer an aspiration—it is the standard.” The presentation emphasized the need for routine next-generation sequencing (NGS) to detect resistance mutations and guide therapy selection. Moreover, safety outcomes were favorable, with diarrhea and rash being the most frequent adverse events, consistent with other EGFR inhibitors.\n",
      "\n",
      "This presentation marked a shift toward comprehensive mutation testing across all lung cancer patients in Spain, supported by new reimbursement policies from the Ministry of Health that incentivize genomic diagnostics in tertiary hospitals.\n"
     ]
    }
   ],
   "source": [
    "random_id = \"4:50a55e0e-0d46-4a3c-85dd-d974565f398f:40\"\n",
    "\n",
    "query = f\"\"\"MATCH (n)\n",
    "WHERE elementId(n) = \"{random_id}\"\n",
    "RETURN n.text AS text\n",
    "\"\"\"\n",
    "\n",
    "with driver.session() as session:\n",
    "    result = session.run(query)\n",
    "    for record in result:\n",
    "        print(record[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d92b7d0b",
   "metadata": {},
   "source": [
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "graphs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
